website_sale_cart_expire/models/website.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2021 Camptocamp (http://www.camptocamp.com).
+# @author Iván Todorovich <ivan.todorovich@gmail.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+from datetime import timedelta
+
+from odoo import _, api, fields, models
+from odoo.osv import expression
+
+_logger = logging.getLogger(__name__)
+
+
+class Website(models.Model):
+    _inherit = "website"
+
+    cart_expire_delay = fields.Float(
+        string="Expire Delay",
+        default=0.0,
+        help="Automatically cancel website orders after the given time.\n"
+        "Set to 0 to disable this feature.",
+    )
+
+    def _get_cart_expire_delay_domain(self):
+        self.ensure_one()
+        expire_date = fields.Datetime.now() - timedelta(hours=self.cart_expire_delay)
+        return [
+            ("website_id", "=", self.id),
+            ("state", "=", "draft"),
+            ("write_date", "<=", expire_date),
+            # We don't want to cancel carts that are already in payment.
+            "|",
+            ("transaction_ids", "=", False),
+            "!",
+            ("transaction_ids.state", "in", ["pending", "authorized", "done"]),
```

Comment:
It's common to see the negative term `"!"` used directly as it's translated by odoo from operator, so I would suggest but won't lock merging process for that 

```suggestion
            ("transaction_ids.state", "not in", ["pending", "authorized", "done"]),
```

Reply:
In this case it's not the same, though.

Using the `!` operator it means: _Orders that don't have any transaction in these states_
Using the `not in` operator it means: _Orders having transactions in states different than these_